### PR TITLE
Refactor socks layer and enhance unittet

### PR DESCRIPTION
**Pull request by chhsiao90**
Had refactor socks layer, please give me feedback if any, thanks!

### CODE REVIEW

**Comment by mike820324** [comment]
Interesting PR, based on the diff it seems like you rewrite the socks layer.

File: microproxy/layer/proxy/socks.py [hunk 1]
```diff
@@ -33,120 +33,108 @@ def process_and_return_context(self):
...
-            response_event = GreetingResponse(
-                socks_version, AUTH_TYPE["NO_AUTH"])
-        return response_event
+            yield self.send_event_to_src_conn(GreetingResponse(
+                socks_version, AUTH_TYPE["NO_AUTH"]))
 
     @gen.coroutine
-    def handle_request(self, event):
+    def handle_request_and_create_destination(self, event):
+        """Handle the socks requet from source
```
Comment (mike820324) [inline microproxy/layer/proxy/socks.py]: requet -> request

File: microproxy/layer/proxy/socks.py [hunk 2]
```diff
@@ -33,120 +33,108 @@ def process_and_return_context(self):
...
+        Create destination connection
 
+        Returns:
+            tuple: (dest_stream, host, port)
+        """
         if event.cmd != REQ_COMMAND["CONNECT"]:
             logger.debug("Unsupport connect type")
-            error = ProtocolError("Unsupport bind type")
-            response_event = Response(
+            self.send_event_to_src_conn(Response(
```
Comment (mike820324) [inline microproxy/layer/proxy/socks.py]: Why there is no yield in here?

File: microproxy/layer/proxy/socks.py [hunk 3]
```diff
@@ -33,120 +33,108 @@ def process_and_return_context(self):
...
-        dest_stream = None
         try:
-            dest_stream = yield self.create_dest_stream((host, port))
-
-        except Exception as e:
-            err, response_event = self.handle_connection_error(
-                e, event, dest_stream)
-
+            dest_stream = yield self.create_dest_stream((event.addr, event.port))
+        except gen.TimeoutError as te:
```
Comment (mike820324) [inline microproxy/layer/proxy/socks.py]: is te and se really necessary? it seems the gen.Timeout and StreamClosedError are pretty self explain.

**Comment by chhsiao90** [comment]
Updated and rebased

The major change of this PR is remove the return of tuple (error, event)
and let the handle_greeting_request , handle_request_and_create_destination to send event and raise Exception.

File: microproxy/layer/proxy/socks.py (hunk 2 above)
Comment (chhsiao90) [reply microproxy/layer/proxy/socks.py]: Sorry, I forgot.

File: microproxy/layer/proxy/socks.py (hunk 3 above)
Comment (chhsiao90) [reply microproxy/layer/proxy/socks.py]: Yeah~

**Comment by mike820324** [comment]
Sorry for taking so long. LGTM now.

**Pull request merged**